File: packages/core/src/detection/pipeline.ts
import { randomUUID } from "node:crypto";

import { analyzeAcceptHeader } from "./signal-accept-header.js";
import { runHeaderHeuristics } from "./signal-header-heuristics.js";
import { checkUaDatabase } from "./signal-ua-database.js";
import { verifyClawdentityToken } from "./verifier-clawdentity.js";
import { verifyRfc9421Signature } from "./verifier-rfc9421.js";

import type { AgentContext, AgentContextBuilder } from "../types/agent-context.js";
import type { AgentRequest } from "../types/agent-request.js";
import type { DetectionConfig } from "../types/config.js";
import type { DetectionSignal, TrustTier } from "../types/trust-tier.js";


/**
 * Layer 0 — Detection Pipeline
 *
 * Runs all 4 detection signals in the optimal order and resolves a TrustTier.
 * Also runs identity verification (RFC 9421 and Clawdentity) for requests that
 * carry signature or token headers.
 *
 * Signal evaluation order (fastest/cheapest first):
 * 1. Accept Header (synchronous string parsing)
 * 2. UA Database (synchronous hash/prefix lookup)
 * 3. Header Heuristics (synchronous header analysis)
 * 4. Identity Verification (async network call — only if signature headers present)
 *
 * Trust Tier Resolution:
 * - Any valid identity verification → "verified-agent"
 * - UA database match → "known-agent"
 * - Header heuristics above threshold → "suspected-agent"
 * - Accept header signal alone (without UA match) → "suspected-agent"
 * - None of the above → "human"
 */

/**
 * Resolve the trust tier from the collected signals and verification results.
 * The tier is always the highest-trust tier earned by any signal.
 */
function resolveTier(
  signals: DetectionSignal[],
  isVerified: boolean,
  isKnown: boolean,
  isSuspected: boolean,
  requestedMarkdown: boolean,
): { tier: TrustTier; reason: string } {
  if (isVerified) {
    return {
      tier: "verified-agent",
      reason: "Cryptographic identity verification passed (RFC 9421 or Clawdentity AIT)",
    };
  }
  if (isKnown) {
    return {
      tier: "known-agent",
      reason: `User-Agent matched database entry (signals: ${signals.join(", ")})`,
    };
  }
  if (isSuspected || (requestedMarkdown && !isKnown)) {
    return {
      tier: "suspected-agent",
      reason: `Heuristic signals indicate agent traffic (signals: ${signals.join(", ")})`,
    };
  }
  return {
    tier: "human",
    reason: "No agent signals detected — treating as human browser",
  };
}

/**
 * Run the full detection pipeline for a single request.
 * Returns a fully-populated AgentContext.
 *
 * This is an async function because identity verification may require a network
 * call to fetch the agent operator's public key JWKS. The JWKS is cached, so
 * only the first request from a new agent domain incurs the latency.
 */
export async function runDetectionPipeline(
  request: AgentRequest,
  config: DetectionConfig = {},
): Promise<AgentContext> {
  const startTime = Date.now();
  const allSignals: DetectionSignal[] = [];

  // -------------------------------------------------------------------------
  // Signal 1: Accept Header (synchronous)
  // -------------------------------------------------------------------------
  const acceptResult = analyzeAcceptHeader(request.headers["accept"]);
  if (acceptResult.hasAgentSignal) {
    allSignals.push(...acceptResult.signals);
  }

  // -------------------------------------------------------------------------
  // Signal 2: UA Database Lookup (synchronous)
  // -------------------------------------------------------------------------
  const uaResult = checkUaDatabase(request.headers["user-agent"], config.customAgents);
  let matchedAgent = null;
  let agentCategory = null;
  if (uaResult.matched && uaResult.match) {
    allSignals.push(...uaResult.signals);
    matchedAgent = uaResult.match.entry;
    agentCategory = uaResult.match.entry.category;
  }

  // -------------------------------------------------------------------------
  // Signal 3: Header Heuristics (synchronous, only if enabled)
  // -------------------------------------------------------------------------
  let heuristicsIsSuspected = false;
  if (config.headerHeuristics !== false) {
    const heuristicsResult = runHeaderHeuristics(request.headers);
    if (heuristicsResult.isSuspected) {
      allSignals.push(...heuristicsResult.signals);
      heuristicsIsSuspected = true;
    }
  }

  // -------------------------------------------------------------------------
  // Signal 4: Identity Verification (async — only if signature/token headers present)
  // -------------------------------------------------------------------------
  let verifiedIdentity = null;
  const hasSignatureHeaders =
    "signature" in request.headers && "signature-input" in request.headers;
  const hasAgentToken = request.headers["authorization"]?.toLowerCase().startsWith("agenttoken");

  if (hasSignatureHeaders) {
    const rfc9421Result = await verifyRfc9421Signature(request);
    if (rfc9421Result.valid && rfc9421Result.identity) {
      allSignals.push("rfc9421-signature");
      verifiedIdentity = rfc9421Result.identity;
    }
  } else if (hasAgentToken && !verifiedIdentity) {
    const clawdentityResult = await verifyClawdentityToken(request.headers["authorization"]);
    if (clawdentityResult.valid && clawdentityResult.identity) {
      allSignals.push("clawdentity-ait");
      verifiedIdentity = clawdentityResult.identity;
    }
  }

  // -------------------------------------------------------------------------
  // Tier Resolution
  // -------------------------------------------------------------------------
  const isKnown = uaResult.matched;
  const isVerified = verifiedIdentity !== null;
  const isSuspected = heuristicsIsSuspected;

  const { tier, reason } = resolveTier(
    allSignals,
    isVerified,
    isKnown,
    isSuspected,
    acceptResult.prefersMarkdown,
  );

  const isAgent = tier !== "human";

  // -------------------------------------------------------------------------
  // Assemble AgentContext
  // -------------------------------------------------------------------------
  const builder: AgentContextBuilder = {
    requestId: randomUUID(),
    receivedAt: new Date().toISOString(),
    tier,
    tierResolution: {
      tier,
      signals: allSignals,
      reason,
    },
    isAgent,
    userAgent: request.headers["user-agent"] ?? "",
    matchedAgent,
    agentCategory,
    signals: allSignals,
    verifiedIdentity,
    tenantContext: null, // Populated by Layer 8 if delegation token is present
    requestedMarkdown: acceptResult.prefersMarkdown,
    path: request.path,
    method: request.method,
    headers: request.headers,
    trace: [
      {
        layer: "Layer0:Detection",
        action: `Resolved tier="${tier}" via signals=[${allSignals.join(",")}]. ${reason}`,
        durationMs: Date.now() - startTime,
      },
    ],
  };

  // AgentContext is sealed as readonly from the outside; the builder is internal only
  return Object.freeze(builder) as AgentContext;
}

File: python_sdk/agentfriendly/detection/pipeline.py
"""
Layer 0 — Detection Pipeline

Orchestrates all 4 signals and resolves a TrustTier for each request.
Mirrors packages/core/src/detection/pipeline.ts
"""

from __future__ import annotations

import time
import uuid
from datetime import datetime, timezone

from ..config import DetectionConfig
from ..types import AgentContext, AgentEntry, TraceEntry, TrustTier, DetectionSignal
from .signal_accept_header import analyze_accept_header
from .signal_header_heuristics import run_header_heuristics
from .signal_ua_database import check_ua_database


def _resolve_tier(
    signals: list[str],
    is_verified: bool,
    is_known: bool,
    is_suspected: bool,
    requested_markdown: bool,
) -> tuple[TrustTier, str]:
    """Resolve the trust tier from collected signals."""
    if is_verified:
        return "verified-agent", "Cryptographic identity verification passed"
    if is_known:
        return "known-agent", f"UA matched database (signals: {', '.join(signals)})"
    if is_suspected or (requested_markdown and not is_known):
        return "suspected-agent", f"Heuristic signals detected (signals: {', '.join(signals)})"
    return "human", "No agent signals detected"


async def run_detection_pipeline(
    *,
    method: str,
    path: str,
    headers: dict[str, str],
    config: DetectionConfig | None = None,
) -> AgentContext:
    """
    Run the full detection pipeline for a single request.
    Returns a fully-populated AgentContext.

    Async because identity verification may require a network call
    to fetch the agent operator's public key (cached after first fetch).
    """
    cfg = config or DetectionConfig()
    start_time = time.monotonic()
    all_signals: list[str] = []

    # -------------------------------------------------------------------------
    # Signal 1: Accept Header
    # -------------------------------------------------------------------------
    accept_result = analyze_accept_header(headers.get("accept"))
    if accept_result.has_agent_signal:
        all_signals.extend(accept_result.signals)

    # -------------------------------------------------------------------------
    # Signal 2: UA Database
    # -------------------------------------------------------------------------
    ua_result = check_ua_database(
        headers.get("user-agent"), cfg.custom_agents or None
    )
    matched_agent: AgentEntry | None = None
    agent_category = None
    if ua_result.matched and ua_result.match:
        all_signals.extend(ua_result.signals)
        matched_agent = ua_result.match.entry
        agent_category = ua_result.match.entry.category

    # -------------------------------------------------------------------------
    # Signal 3: Header Heuristics
    # -------------------------------------------------------------------------
    heuristics_is_suspected = False
    if cfg.header_heuristics:
        heuristics_result = run_header_heuristics(headers)
        if heuristics_result.is_suspected:
            all_signals.extend(heuristics_result.signals)
            heuristics_is_suspected = True

    # -------------------------------------------------------------------------
    # Signal 4: Identity Verification (RFC 9421 / Clawdentity)
    # Currently Python-side performs structural verification only.
    # Full async Ed25519 verification is included when cryptography is installed.
    # -------------------------------------------------------------------------
    verified_identity = None
    has_sig_headers = "signature" in headers and "signature-input" in headers
    has_agent_token = (headers.get("authorization", "")).lower().startswith("agenttoken")

    if has_sig_headers:
        try:
            from .verifier_rfc9421 import verify_rfc9421_signature

            result = await verify_rfc9421_signature(
                method=method,
                url=f"https://placeholder{path}",
                headers=headers,
            )
            if result.valid and result.identity:
                all_signals.append("rfc9421-signature")
                verified_identity = result.identity
        except ImportError:
            pass  # cryptography package not installed

    elif has_agent_token and not verified_identity:
        try:
            from .verifier_clawdentity import verify_clawdentity_token

            result = await verify_clawdentity_token(headers.get("authorization"))
            if result.valid and result.identity:
                all_signals.append("clawdentity-ait")
                verified_identity = result.identity
        except ImportError:
            pass

    # -------------------------------------------------------------------------
    # Tier Resolution
    # -------------------------------------------------------------------------
    tier, reason = _resolve_tier(
        all_signals,
        is_verified=verified_identity is not None,
        is_known=ua_result.matched,
        is_suspected=heuristics_is_suspected,
        requested_markdown=accept_result.prefers_markdown,
    )

    is_agent = tier != "human"
    duration_ms = (time.monotonic() - start_time) * 1000

    context = AgentContext(
        request_id=str(uuid.uuid4()),
        received_at=datetime.now(timezone.utc).isoformat(),
        tier=tier,
        signals=list(all_signals),  # type: ignore[arg-type]
        is_agent=is_agent,
        user_agent=headers.get("user-agent", ""),
        matched_agent=matched_agent,
        agent_category=agent_category,
        verified_identity=verified_identity,
        tenant_context=None,
        requested_markdown=accept_result.prefers_markdown,
        path=path,
        method=method.upper(),
        headers={k.lower(): v for k, v in headers.items()},
        trace=[
            TraceEntry(
                layer="Layer0:Detection",
                action=f'Resolved tier="{tier}" via signals=[{", ".join(all_signals)}]. {reason}',
                duration_ms=duration_ms,
            )
        ],
    )

    return context
